parsec/core/backend_connection/transport.py

```diff
@@ -114,7 +127,280 @@ async def _connect(
     return transport
 
 
-def _upgrade_stream_to_ssl(raw_stream, hostname):
+def cook_basic_auth_header(username: str, password: str) -> str:
+    step1 = f"{username}:{password}".encode("utf8")
+    step2 = b64encode(step1).decode("ascii")
+    return f"Basic {step2}"
+
+
+# Global vars to easily set up proxy from config file or cli params
+
+
+__force_proxy_pac_url = None
+
+
+def force_proxy_pac_url(url: str) -> None:
+    global __force_proxy_pac_url
+    __force_proxy_pac_url = url
+
+
+__force_proxy_url = None
+
+
+def force_proxy_url(url: str) -> None:
+    global __force_proxy_url
+    __force_proxy_url = url
+
+
+def _build_http_url(hostname: str, port: int, use_ssl: bool) -> str:
+    if use_ssl:
+        url = f"https://{hostname}"
+        if port != 443:
+            url += f":{port}"
+    else:
+        url = f"http://{hostname}"
+        if port != 80:
+            url += f":{port}"
+    return url
+
+
+# def _get_proxy_from_pac_config(hostname: str, port: int, use_ssl: bool) -> Optional[str]:
+def _get_proxy_from_pac_config(url: str, hostname: str) -> Optional[str]:
+    # Hack to disable check on content-type (given server migh not be well
+    # configured, and `get_pac` silently fails on wrong content-type by returing None)
+    # Also disable WPAD protocol to retrieve PAC from DNS given it produce annoying
+    # WARNING logs
+    get_pac_kwargs: dict = {"from_dns": False, "allowed_content_types": {""}}
+    if __force_proxy_pac_url:
+        get_pac_kwargs["url"] = __force_proxy_pac_url
+        logger.debug(
+            "Retreiving .PAC proxy config url from forced location", pac_url=__force_proxy_pac_url
+        )
+    elif __force_proxy_url:
+        # Don't try to superstep the forced config !
+        return None
+    else:
+        logger.debug("Retreiving .PAC proxy config url from system")
+    try:
+        pacfile = pypac.get_pac(**get_pac_kwargs)
+    except Exception as exc:
+        logger.warning("Error while retrieving .PAC proxy config", exc_info=exc)
+        return None
+
+    if not pacfile:
+        return None
+
+    try:
+        proxies = pacfile.find_proxy_for_url(url, hostname)
+        logger.debug("Found proxies info in .PAC proxy config", target_url=url, proxies=proxies)
+        proxies = [p.strip() for p in proxies.split(";")]
+        if len(proxies) > 1:
+            logger.warning(
+                "Found multiple proxies info in .PAC proxy config for target, only the first one is going to be used !",
+                target_url=url,
+                proxies=proxies,
+            )
+        # We don't handle multiple proxies so keep the first correct one and pray !
+        for proxy in proxies:
+            if proxy in ("DIRECT", ""):
+                # PAC explicitly told us not to use a proxy
+                return ""
```

Comment:
The special use of `""` as an explicit way to not use proxy feels weird, can we have a dedicated value for it? And maybe something that does not evaluate to `False` in order to not be confused with `None`?

Reply:
you you mean the returned `""` value vs `None`, I've made this more explicit by adding docstring in `_get_proxy_*` and having `get_proxy` getting rid of it (hence only `get_proxy` call `_get_proxy_*` and have to deal with the `""` special case)